simulation/sub8_simulation/sub8_sim_tools/physics/vehicle.py

```diff
@@ -121,6 +171,62 @@ def publish_pose(self):
 
         self.truth_odom_pub.publish(odom_msg)
 
+    def publish_imu(self):
+        """Publishes imu sensor information - orientation, angular velocity, and linear acceleration"""
+        pose_matrix = np.transpose(self.pose)
+        linear_vel = self.body.getRelPointVel(self.dvl_position)
+        angular_vel = self.body.getAngularVel()
+        quaternion = tf.transformations.quaternion_from_matrix(pose_matrix)
+       
+        header = Header(
+            stamp=rospy.Time.now(),
+            frame_id='/world'
+        )
+
+        linear=geometry.Vector3(*linear_vel)
+        angular=geometry.Vector3(*angular_vel)
+
+        noise = np.random.random(3)
+
+
+        covariance = [noise[0], 0., 0.,
```

Comment:
You appear not to know what covariance is... It's a description of the noise present, not the actual noise signal.

You should be adding `noise` to one of the actual measured quantities (`angular` or `linear`), not using it as the covariance. The covariance should be a constant diagonal matrix.

Also, an accelerometer measures acceleration, not linear velocity.

I wouldn't bother filling in the orientation. An IMU doesn't directly measure orientation, so it's kind of weird that it's in the IMU message at all. The old software didn't fill it in or use it anywhere. [The message](http://docs.ros.org/api/sensor_msgs/html/msg/Imu.html) says

>If you have no estimate for one of the data elements (e.g. your IMU doesn't produce an orientation estimate), please set element 0 of the associated covariance matrix to -1